scanner/src/main/kotlin/ScannerResultBuilder.kt

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (C) 2020 Bosch.IO GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner
+
+import java.time.Instant
+
+import org.ossreviewtoolkit.model.AccessStatistics
+import org.ossreviewtoolkit.model.Environment
+import org.ossreviewtoolkit.model.OrtResult
+import org.ossreviewtoolkit.model.ScanRecord
+import org.ossreviewtoolkit.model.ScanResultContainer
+import org.ossreviewtoolkit.model.ScannerRun
+import org.ossreviewtoolkit.model.config.ScannerConfiguration
+
+/**
+ * Definition of an interface that allows a scanner to construct an [OrtResult] with a scanner result without
+ * having to know about details.
+ *
+ * This interface supports a more efficient handling of large scanner results. The basic idea is that a scanner
+ * adds results for single packages when they become available and does not have to care how these results are
+ * combined to a full [OrtResult].
+ *
+ * Concrete implementations handle the result construction differently: A naive implementation can collect all
+ * results in memory, while a more efficient one streams the data to a file reducing the memory consumption.
+ */
+interface ScannerResultBuilder : AutoCloseable {
+    /**
+     * Initialize this builder from the [OrtResult][analyzerResult] produced by the analyzer. The result produced by
+     * this builder is then based on the result passed to this function. This function must be called initially.
+     */
+    fun initFromAnalyzerResult(analyzerResult: OrtResult)
+
+    /**
+     * Add a [resultContainer] to the result managed by this builder. This function should be called when scan
+     * results for a package become available.
+     */
+    fun addScanResult(resultContainer: ScanResultContainer)
```

Comment:
Hm, I think, the way the scanning logic works, this cannot happen: A builder is confined to a single scanner, and the scanner produces only a single result container for a package. But this could be stated more explicitly in the comment.